url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/Keyboard.java

```diff
@@ -0,0 +1,619 @@
+
+    if (hotkeyScript != null) {
+      int mod = event.getModifiersEx();
+//    System.out.println("HotKey: " + key + " mod: " + mod);
+
+      for (int i = 0, n = hotkeyScript.size(); i < n; i++) {
+        Object[] hk = (Object[]) hotkeyScript.get(i);
+        int[] keys = (int[]) hk[2];
+//      System.out.println("Cmp " + keys[0] + " mod: " + keys[1]);
+        if (keys[0] == key && ((mod & keys[1]) == keys[1])) {
+          c64script.interpretCall((String) hk[0], hk[1]);
+        }
+      }
+    }
+
+    if (key == 0) {
+      System.out.println("KeyZero ???");
+      key = (int) Character.toLowerCase(chr);
+    }
+
+    if (key == restoreKey) {
+      screen.restoreKey(true);
+    }
+
+    if (key != lastKey) {
+      keyPressed++;
+    }
+    lastKey = key;
+
+    int usr = getUserStick(key, location);
+    if (extendedKeyboardEmulation) usr = -1;
+    if (usr == -1) {
+      usr = getNormalStick(key);
+    }
+
+    switch (usr) {
+    // CHANGE TO STOCK1LEFT = true; (or similar!)
+    case USER_UP:
+      joystick1 = joystick1 & (255 - STICK_UP);
+      lastUp = true;
+      updateJoystick();
+      if (stickExits) {
+        return;
+      }
+      break;
+    case USER_DOWN:
+      joystick1 = joystick1 & (255 - STICK_DOWN);
+      lastUp = false;
+      updateJoystick();
+      if (stickExits) {
+        return;
+      }
+      break;
+    case USER_LEFT:
+      joystick1 = joystick1 & (255 - STICK_LEFT);
+      lastLeft = true;
+      updateJoystick();
+      if (stickExits) {
+        return;
+      }
+      break;
+    case USER_RIGHT:
+      joystick1 = joystick1 & (255 - STICK_RIGHT);
+      lastLeft = false;
+      updateJoystick();
+      if (stickExits) {
+        return;
+      }
+      break;
+    case USER_FIRE:
+      joystick1 = joystick1 & (255 - STICK_FIRE);
+      updateJoystick();
+      if (stickExits) {
+        return;
+      }
+      break;
+    }
+
+    // Change the joystick with F10!
+    switch (key) {
+    case KeyEvent.VK_F10:
+      setStick(stick == IO_OFFSET + 0xdc00);
+      break;
+    case KeyEvent.VK_F9:
+      System.out.println("F9");
+      extendedKeyboardEmulation = !extendedKeyboardEmulation;
+      stickExits = !extendedKeyboardEmulation;
+      break;
+    case KeyEvent.VK_F11:
+      // Toggle colour set?
+      break;
+    }
+
+    if (extendedKeyboardEmulation) {
+      // Auto shift handling!
+      if ((keytable[key][2] & AUTO_SHIFT) != 0) {
+        keyShift++;
+        if (keyShift == 1) {
+          // AutoShift on if first shift! - otherwize not?!
+          handleKeyPress(KeyEvent.VK_SHIFT, location);
+        }
+        // Also increase shift is shift is pressed
+      } else if (key == KeyEvent.VK_SHIFT) {
+        keyShift++;
+      }
+      handleKeyPress(key, location);
+
+    } else {
+      // No autoshift keys!!!
+      if (keytable[key][2] < MIN_AUTO) {
+        handleKeyPress(key, location);
+      }
+    }
+  }
+
+  public void keyReleased(KeyEvent event)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 487-507 into a separate method.